src/main/java/com/snowflake/core/util/HiveSyncTool.java

```diff
@@ -0,0 +1,232 @@
+
+  // Note: HiveMetaStoreClient is not backwards compatible. Notably, 2.X
+  //       versions use a HiveConf argument instead of a Configuration
+  //       argument. Also, if the version of the HiveMetaStoreClient is not
+  //       compatible with the metastore, calls to the metastore may return
+  //       wrong results, e.g. client.getAllDatabases() returns no databases.
+  private final HiveMetaStoreClient hmsClient;
+
+  // The Snowflake configuration
+  private final SnowflakeConf snowflakeConf;
+
+  /**
+   * Instantiates the HiveSyncTool.
+   * @param hmsClient a client for the Hive metastore
+   */
+  public HiveSyncTool(HiveMetaStoreClient hmsClient)
+  {
+    this.hmsClient = hmsClient;
+    this.snowflakeConf = new SnowflakeConf();
+  }
+
+  /**
+   * Does a one-time, one-way metadata sync from the Hive metastore to
+   * Snowflake.
+   * 1. Tables in Hive are created in Snowflake
+   * 2. Partitions not in Hive are dropped from Snowflake
+   * 3. Partitions in Hive are added to Snowflake
+   * Note: does not drop tables from Snowflake.
+   * @throws TException thrown when encountering a Thrift exception while
+   *         communicating with the metastore or executing a metastore operation
+   */
+  public void sync() throws TException
+  {
+    Pattern tableNameFilter = snowflakeConf.getPattern(
+        SnowflakeConf.ConfVars.SNOWFLAKE_TABLE_FILTER_REGEX.getVarname(), null);
+    Pattern databaseNameFilter = snowflakeConf.getPattern(
+        SnowflakeConf.ConfVars.SNOWFLAKE_DATABASE_FILTER_REGEX.getVarname(), null);
+
+    log.info("Starting sync");
+    List<String> databaseNames = hmsClient.getAllDatabases().stream()
+        .filter(db -> databaseNameFilter == null || !databaseNameFilter.matcher(db).matches())
+        .collect(Collectors.toList());
+    log.info(String.format("Syncing %s databases from Hive",
+                           databaseNames.size()));
+    for (String databaseName : databaseNames)
+    {
+      Preconditions.checkNotNull(databaseName);
+      List<String> tableNames = hmsClient.getAllTables(databaseName).stream()
+          .filter(table -> tableNameFilter == null || !tableNameFilter.matcher(table).matches())
+          .collect(Collectors.toList());
+      log.info(String.format("Syncing %s tables for database %s",
+                             tableNames.size(), databaseName));
+      for (String tableName : tableNames)
+      {
+        Preconditions.checkNotNull(tableName);
+
+        // Do a touch on the table to fire listener events
+        Table hiveTable = hmsClient.getTable(databaseName, tableName);
+        hmsClient.alter_table(databaseName, tableName, hiveTable);
+
+        if (!hiveTable.getPartitionKeys().isEmpty())
+        {
+          // Drop extra partitions
+          dropExtraPartitionsFromSnowflake(databaseName, hiveTable);
+        }
+
+        // Do a touch on all partitions to fire listener events
+        List<Partition> partitions = hmsClient.listPartitions(
```

Comment:
does hive metastore client have an iterable API for listPartitions? A table can have several thousand partitions - potentially causing issues on the hive metastore

Reply:
I don't think it has one, at least not in the version we're using (or even relatively newer versions). However, we could use listPartitionValues, get a PartitionValuesResponse (that supports getPartitionValuesIterator()), and get each partition individually. Alternatively, we could list partition names (and get a List<String>) and get the real partitions in chunks.

But I didn't see a way to page results of partitions or partition names, nor a way to convert a batch of partition values (which should be paged) to partitions or partition names.